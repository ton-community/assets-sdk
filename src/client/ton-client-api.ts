import { Address, Cell, Contract, ContractProvider, OpenedContract } from '@ton/core';
import { TonClient4 } from '@ton/ton';
import { getHttpV4Endpoint } from '@orbs-network/ton-access';

/**
 * The TonClientApi interface provides a way to interact with the TON network
 */
export interface TonClientApi {
    /**
     * Open a contract
     * @param contract - The contract to open
     */
    open<T extends Contract>(contract: T): OpenedContract<T>;

    /**
     * Create a new contract provider
     * @param address - The address of the contract
     * @param init - The initial state of the contract
     */
    provider(address: Address, init?: { code: Cell; data: Cell } | null | undefined): ContractProvider;
}

/**
 * Create a new TonClient4 instance with the given network
 * @param network - The network to connect to (testnet or mainnet)
 */
export async function createApi(network: 'testnet' | 'mainnet'): Promise<TonClientApi> {
    const endpoint = await getHttpV4Endpoint({ network: network });
    return new TonClient4({ endpoint: endpoint, timeout: 15000 });
}
