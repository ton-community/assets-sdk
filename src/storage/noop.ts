import { Storage } from './storage';

export class NoopStorage implements Storage {
    async uploadFile(_contents: Buffer): Promise<string> {
        throw new Error(`No storage provider configured`);
    }
}
