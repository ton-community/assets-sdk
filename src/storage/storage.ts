export interface Storage {
    uploadFile(contents: Buffer): Promise<string>;
}
