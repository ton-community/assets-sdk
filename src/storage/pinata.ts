import type { Readable } from 'stream';

import type PinataClient from '@pinata/sdk';

import { defer, Deferred } from '../utils';
import { Storage } from './storage';

export interface PinataStorageParams {
    pinataApiKey: string;
    pinataSecretKey: string;
}

export class PinataStorage implements Storage {
    private readonly apiKey: string;

    private readonly secretApiKey: string;

    private readonly client: Deferred<PinataClient> = defer(async () => {
        const pinata = await import('@pinata/sdk').then((m) => m.default);
        return new pinata(this.apiKey, this.secretApiKey);
    });

    private readonly stream: Deferred<Readable, [Buffer]> = defer(async (contents: Buffer) => {
        const stream = await import('stream').then((m) => m.Readable);
        return stream.from(contents);
    });

    constructor(apiKey: string, secretApiKey: string) {
        this.apiKey = apiKey;
        this.secretApiKey = secretApiKey;
    }

    public static create(params: PinataStorageParams) {
        return new PinataStorage(params.pinataApiKey, params.pinataSecretKey);
    }

    async uploadFile(contents: Buffer): Promise<string> {
        const client = await this.client();
        const stream = await this.stream(contents);

        const result = await client.pinFileToIPFS(stream, {
            pinataMetadata: {
                name: 'Assets SDK Jetton',
            },
        });
        return 'ipfs://' + result.IpfsHash;
    }
}
