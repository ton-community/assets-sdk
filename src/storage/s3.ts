import type { S3 } from '@aws-sdk/client-s3';

import { defer, Deferred } from '../utils';
import { Storage } from './storage';

export interface S3StorageParams {
    s3AccessKeyId: string;
    s3SecretAccessKey: string;
    s3Bucket: string;
}

export class S3Storage implements Storage {
    private readonly accessKeyId: string;

    private readonly secretAccessKey: string;

    private readonly bucket: string;

    private readonly s3: Deferred<S3> = defer(async () => {
        const s3 = await import('@aws-sdk/client-s3').then((m) => m.S3);
        return new s3({
            credentials: {
                accessKeyId: this.accessKeyId,
                secretAccessKey: this.secretAccessKey,
            },
        });
    });

    constructor(accessKeyId: string, secretAccessKey: string, bucket: string) {
        this.accessKeyId = accessKeyId;
        this.secretAccessKey = secretAccessKey;
        this.bucket = bucket;
    }

    public static create(params: S3StorageParams) {
        return new S3Storage(params.s3AccessKeyId, params.s3SecretAccessKey, params.s3Bucket);
    }

    async uploadFile(contents: Buffer): Promise<string> {
        const s3 = await this.s3();

        const key = 'jetton/' + Math.random().toString(36).substring(2);

        await s3.putObject({
            Bucket: this.bucket,
            Key: key,
            Body: contents,
        });

        return 'https://' + this.bucket + '.s3.amazonaws.com/' + key;
    }
}
