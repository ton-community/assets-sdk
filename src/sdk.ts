import { Address, beginCell, Sender } from '@ton/core';

import { Storage } from './storage/storage';
import { PinataStorage, PinataStorageParams } from './storage/pinata';
import { S3Storage, S3StorageParams } from './storage/s3';
import { TonClientApi } from './client/ton-client-api';
import { JettonContent, jettonContentToInternal } from './jetton/content';
import { NftContent, nftContentToInternal } from './nft/content';
import { internalOnchainContentToCell } from './utils';
import { JettonWallet } from './jetton/JettonWallet';
import { JettonMinter } from './jetton/JettonMinter';
import { NftCollection } from './nft/NftCollection';
import { NftItem } from './nft/NftItem';
import { SbtCollection } from './nft/SbtCollection';
import { NftSaleParams } from './nft/data';
import { ContentResolver, DefaultContentResolver } from './content';
import { NftSale } from './nft/NftSale';
import { NoopStorage } from './storage/noop';
import { SendTransferOptions } from './common/types';
import { NftRoyaltyParams } from './nft/types/NftRoyaltyParams';
import { NftMintItemParams } from './nft/types/NftBatchMintMessage';
import { SbtItemParams } from './nft/types/SbtItemParams';
import { NftItemParams } from './nft/types/NftItemParams';
import { retry } from './cli/common';

const WORKCHAIN = 0;

type DeployJettonOptions = {
    onchainContent?: boolean;
    adminAddress?: Address;
    premintAmount?: bigint;
    premintOptions?: SendTransferOptions;
    value?: bigint;
    queryId?: bigint;
};

export class AssetsSDK {
    constructor(
        public readonly storage: Storage,
        public readonly api: TonClientApi,
        public readonly sender?: Sender,
        public readonly contentResolver?: ContentResolver,
    ) {}

    static create(params: {
        api: TonClientApi;
        storage?: PinataStorageParams | S3StorageParams | Storage;
        sender?: Sender;
        contentResolver?: ContentResolver;
    }) {
        let { api, storage, sender: sender, contentResolver } = params;

        if (!storage) {
            storage = new NoopStorage();
        } else if ('pinataApiKey' in storage) {
            storage = PinataStorage.create(storage);
        } else if ('s3AccessKeyId' in storage) {
            storage = S3Storage.create(storage);
        }

        contentResolver ??= new DefaultContentResolver();

        return new AssetsSDK(storage, api, sender, contentResolver);
    }

    async deployJetton(content: JettonContent, options?: DeployJettonOptions) {
        if (!this.sender) {
            throw new Error('Sender must be defined');
        }

        const adminAddress = options?.adminAddress ?? this.sender?.address;
        if (adminAddress === undefined) {
            throw new Error('Admin address must be defined in options or be available in Sender');
        }

        const jettonMinterContract = JettonMinter.createFromConfig(
            {
                admin: adminAddress,
                content: await this.contentToCell(jettonContentToInternal(content), options?.onchainContent ?? false),
            },
            JettonMinter.code,
            WORKCHAIN,
            this.contentResolver,
        );
        const jetton = this.api.open(jettonMinterContract);

        const premintAmount = options?.premintAmount;

        if (typeof premintAmount === 'bigint' && premintAmount > 0n) {
            await jetton.sendMint(this.sender, adminAddress, premintAmount, {
                ...options?.premintOptions,
                value: options?.value,
                queryId: options?.queryId,
            });
        } else {
            await jetton.sendDeploy(this.sender, options?.value);
        }
        return jetton;
    }

    openJetton(address: Address) {
        return this.api.open(JettonMinter.createFromAddress(address, this.contentResolver));
    }

    async deployNftCollection(
        content: { collectionContent: NftContent; commonContent: string },
        options?: {
            royaltyParams?: NftRoyaltyParams;
            adminAddress?: Address;
            onchainContent?: boolean;
            premintItems?: NftMintItemParams<NftItemParams>[];
            value?: bigint;
            queryId?: bigint;
        },
    ) {
        if (!this.sender) {
            throw new Error('Sender must be defined');
        }

        const adminAddress = options?.adminAddress ?? this.sender?.address;
        if (adminAddress === undefined) {
            throw new Error('Admin address must be defined in options or be available in Sender');
        }

        const collection = this.api.open(
            NftCollection.createFromConfig(
                {
                    admin: adminAddress,
                    content: beginCell()
                        .storeRef(
                            await this.contentToCell(
                                nftContentToInternal(content.collectionContent),
                                options?.onchainContent ?? false,
                            ),
                        )
                        .storeRef(beginCell().storeStringTail(content.commonContent))
                        .endCell(),
                    royalty: options?.royaltyParams,
                },
                NftCollection.code,
                WORKCHAIN,
                this.contentResolver,
            ),
        );

        if (typeof options?.premintItems?.length === 'number' && options?.premintItems.length > 0) {
            await collection.sendBatchMint(this.sender, options?.premintItems, {
                value: options?.value,
                queryId: options?.queryId,
            });
        } else {
            await collection.sendDeploy(this.sender, options?.value);
        }

        return collection;
    }

    openNftCollection(address: Address) {
        return this.api.open(NftCollection.createFromAddress(address, this.contentResolver));
    }

    async deploySbtCollection(
        content: {
            collectionContent: NftContent;
            commonContent: string;
            onchainContent?: boolean;
        },
        options?: {
            adminAddress?: Address;
            premintItems?: NftMintItemParams<SbtItemParams>[];
            value?: bigint;
            queryId?: bigint;
        },
    ) {
        if (!this.sender) {
            throw new Error('Sender must be defined');
        }

        const adminAddress = options?.adminAddress ?? this.sender?.address;
        if (adminAddress === undefined) {
            throw new Error('Admin address must be defined in options or be available in Sender');
        }

        const collection = this.api.open(
            SbtCollection.createFromConfig(
                {
                    admin: adminAddress,
                    content: beginCell()
                        .storeRef(
                            await this.contentToCell(
                                nftContentToInternal(content.collectionContent),
                                content?.onchainContent ?? false,
                            ),
                        )
                        .storeRef(beginCell().storeStringTail(content.commonContent))
                        .endCell(),
                },
                SbtCollection.code,
                WORKCHAIN,
                this.contentResolver,
            ),
        );

        if (typeof options?.premintItems?.length === 'number' && options?.premintItems.length > 0) {
            await collection.sendBatchMint(this.sender, options?.premintItems, {
                value: options?.value,
                queryId: options?.queryId,
            });
        } else {
            await collection.sendDeploy(this.sender, options?.value);
        }

        return collection;
    }

    openSbtCollection(address: Address) {
        return this.api.open(SbtCollection.createFromAddress(address, this.contentResolver));
    }

    openJettonWallet(address: Address) {
        return this.api.open(new JettonWallet(address));
    }

    openNftItem(address: Address) {
        return this.api.open(new NftItem(address, undefined, this.contentResolver));
    }

    async deployNftSale(params: NftSaleParams) {
        if (!this.sender) {
            throw new Error('Sender must be defined');
        }

        const marketplaceAddress = params.marketplace ?? this.sender?.address;
        if (marketplaceAddress === undefined) {
            throw new Error('Marketplace address must be defined in options or be available in Sender');
        }
        const sale = this.api.open(
            NftSale.createFromConfig({
                createdAt: params.createdAt ?? Math.floor(Date.now() / 1000),
                marketplace: params.marketplace ?? null,
                nft: params.nft,
                fullPrice: params.fullPrice,
                marketplaceFeeTo: params.marketplaceFeeTo ?? null,
                marketplaceFee: params.marketplaceFee ?? 0n,
                royaltyTo: params.royaltyTo ?? null,
                royalty: params.royalty ?? 0n,
                canDeployByExternal: params.canDeployByExternal ?? true,
            }),
        );
        await sale.sendTopup(this.sender, { value: params.value, queryId: params.queryId });
        return sale;
    }

    openNftSale(address: Address) {
        return this.api.open(NftSale.createFromAddress(address));
    }

    private async internalOffchainContentToCell(internal: Record<string, string | number | undefined>) {
        const contents = Buffer.from(JSON.stringify(internal), 'utf-8');
        const contentUrl = await retry(() => this.storage.uploadFile(contents), { name: 'upload content' });
        return beginCell().storeUint(0x01, 8).storeStringTail(contentUrl).endCell();
    }

    private async contentToCell(internal: Record<string, string | number | undefined>, onchain: boolean) {
        return onchain ? internalOnchainContentToCell(internal) : await this.internalOffchainContentToCell(internal);
    }
}
