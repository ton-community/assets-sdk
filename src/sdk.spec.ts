import '@ton/test-utils';
import { Address, Cell, Contract, ContractProvider, OpenedContract, toNano } from '@ton/core';
import { Blockchain, createMetricStore, makeSnapshotMetric, SandboxContract, TreasuryContract } from '@ton/sandbox';
import { resetMetricStore } from '@ton/sandbox';

import { AssetsSDK } from './sdk';
import { TonClientApi } from './client/ton-client-api';
import { Storage } from './storage/storage';

class Api implements TonClientApi {
    constructor(readonly blockchain: Blockchain) {}
    open<T extends Contract>(contract: T): OpenedContract<T> {
        return this.blockchain.openContract<T>(contract) as OpenedContract<T>;
    }
    provider(address: Address, init?: { code: Cell; data: Cell } | null | undefined): ContractProvider {
        return this.blockchain.provider(address, init);
    }
}

export class FakeStorage implements Storage {
    async uploadFile(_contents: Buffer): Promise<string> {
        return 'FakeStorage';
    }
}

describe('sdk', () => {
    const store = createMetricStore();
    let sdk: AssetsSDK;

    beforeAll(async () => {
        const blockchain = await Blockchain.create();
        const treasury: SandboxContract<TreasuryContract> = await blockchain.treasury('sender')
        const api = new Api(blockchain);
        const storage = new FakeStorage();
        const sender = blockchain.sender(treasury.address);
        sdk = AssetsSDK.create({
            api,
            storage,
            sender,
        });
    });

    afterEach(() => {
        resetMetricStore();
    });

    it('deploy Jetton', async () => {
        await sdk.deployJetton(
            {
                name: 'Jetton',
                decimals: 9,
                symbol: 'JETTON',
            },
            {
                premintAmount: toNano('100'),
            },
        );
        const snapshot = makeSnapshotMetric(store);
        expect(snapshot.items).toMatchSnapshot();
    });

    it('deploy NFT', async () => {
        await sdk.deployNftCollection({
            collectionContent: {
                name: 'NFT',
            },
            commonContent: 'content',
        });
        const snapshot = makeSnapshotMetric(store);
        expect(snapshot.items).toMatchSnapshot();
    });
});
