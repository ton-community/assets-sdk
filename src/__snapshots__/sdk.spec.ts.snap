// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sdk deploy Jetton 1`] = `
[
  {
    "address": "EQCfYtnF6GYZs76Cmrtp8DE4TKR3EPy8ufV_m_IbJHBplhRU",
    "codeHash": "0x0571976c63ec1b7550230a2609dbedb36e1b64ef8d022a16b34ea57063185b2f",
    "contractName": "JettonMinter",
    "execute": {
      "action": undefined,
      "compute": {
        "type": "skipped",
      },
    },
    "message": {
      "in": {
        "bits": 1783,
        "cells": 3,
      },
      "out": {
        "bits": 1000,
        "cells": 2,
      },
    },
    "methodName": "sendMint",
    "opCode": "0x15",
    "receiver": "internal",
    "state": {
      "code": {
        "bits": 5046,
        "cells": 14,
      },
      "data": {
        "bits": 6445,
        "cells": 20,
      },
    },
    "testName": "sdk deploy Jetton",
  },
  {
    "address": "EQBPABJHLy9WThhpL5UIiDIrUHWzz6Mjhq96hPP4TuMkGPsk",
    "codeHash": "0x0571976c63ec1b7550230a2609dbedb36e1b64ef8d022a16b34ea57063185b2f",
    "contractName": "JettonMinter",
    "execute": {
      "action": {
        "resultCode": 0,
        "skippedActions": 0,
        "success": true,
        "totalActionFees": 0,
        "totalActions": 0,
        "totalFwdFees": undefined,
        "totalMessageSize": {
          "bits": 0,
          "cells": 0,
        },
      },
      "compute": {
        "exitCode": 0,
        "gasUsed": 309,
        "success": true,
        "type": "vm",
        "vmSteps": 5,
      },
    },
    "message": {
      "in": {
        "bits": 993,
        "cells": 1,
      },
      "out": {
        "bits": 0,
        "cells": 0,
      },
    },
    "methodName": "notSupported",
    "opCode": "0xffffffff",
    "receiver": "internal",
    "state": {
      "code": {
        "bits": 5046,
        "cells": 14,
      },
      "data": {
        "bits": 6445,
        "cells": 20,
      },
    },
    "testName": "sdk deploy Jetton",
  },
]
`;

exports[`sdk deploy NFT 1`] = `
[
  {
    "address": "EQCERHO2jMd_eP36U6WjsoOvy0zWkdY0FKWhqtDbd1z_tqcs",
    "codeHash": "0x64bb2d4661b5f2dc1a83bf5cbbe09e92ac0b460a1b879a5519386fca4c348bca",
    "contractName": "NftCollection",
    "execute": {
      "action": undefined,
      "compute": {
        "type": "skipped",
      },
    },
    "message": {
      "in": {
        "bits": 681,
        "cells": 1,
      },
      "out": {
        "bits": 744,
        "cells": 2,
      },
    },
    "methodName": "sendDeploy",
    "opCode": "0x0",
    "receiver": "internal",
    "state": {
      "code": {
        "bits": 3804,
        "cells": 20,
      },
      "data": {
        "bits": 4223,
        "cells": 19,
      },
    },
    "testName": "sdk deploy NFT",
  },
  {
    "address": "EQBPABJHLy9WThhpL5UIiDIrUHWzz6Mjhq96hPP4TuMkGPsk",
    "codeHash": "0x64bb2d4661b5f2dc1a83bf5cbbe09e92ac0b460a1b879a5519386fca4c348bca",
    "contractName": "NftCollection",
    "execute": {
      "action": {
        "resultCode": 0,
        "skippedActions": 0,
        "success": true,
        "totalActionFees": 0,
        "totalActions": 0,
        "totalFwdFees": undefined,
        "totalMessageSize": {
          "bits": 0,
          "cells": 0,
        },
      },
      "compute": {
        "exitCode": 0,
        "gasUsed": 309,
        "success": true,
        "type": "vm",
        "vmSteps": 5,
      },
    },
    "message": {
      "in": {
        "bits": 737,
        "cells": 1,
      },
      "out": {
        "bits": 0,
        "cells": 0,
      },
    },
    "methodName": "notSupported",
    "opCode": "0xffffffff",
    "receiver": "internal",
    "state": {
      "code": {
        "bits": 3804,
        "cells": 20,
      },
      "data": {
        "bits": 4223,
        "cells": 19,
      },
    },
    "testName": "sdk deploy NFT",
  },
]
`;
